design-proposals/nic-hotplug/nic-hotplug.md

```diff
@@ -0,0 +1,142 @@
+standard available in multiple platforms.
+This would help bridge the gap between KubeVirt's feature set and what the user
+base expects from it.
+
+## Goals
+Adding network interfaces to running VMs.
+Removing networking interfaces from running VMs.
+A VM can have multiple interfaces connected to the same (secondary) network(s).
+
+## Non Goals
+Providing multiple connections to the primary cluster network (managed by
+Kubernetes).
+
+## Definition of Users
+This feature is intended for cluster users who want to connect their existing
+Virtual Machines to additional existing networks.
+These users may - or may not - have permissions to create the aforementioned
+additional networks; if so, they would need to rely on the cluster
+administrator to provision it for them.
+
+## User Stories
+* as a user, I want to add a new network interface to a running VM.
+* as a user, I want to remove a new network interface from a running VM.
+
+## Repos
+- [CNAO](https://github.com/kubevirt/cluster-network-addons-operator)
+- [KubeVirt](https://github.com/kubevirt/kubevirt)
+- [Multus](https://github.com/k8snetworkplumbingwg/multus-cni)
+
+# Design
+
+## Multus
+Multus - a CNI plugin - only handles the ADD / REMOVE verb, and is triggered
+by kubelet only when the pod's sandbox is created - or removed. Given its
+simplicity, it assumes no networks exist whenever it is executed, and procceeds
+to call the ADD / DEL for **all** networks listed in its
+`k8s.v1.cni.cncf.io/networks` annotation.
+
+As such, the first part of the design implementation must focus on multus; it
+must be refactored to work as a thick plugin, which enables it to be triggered
+not only when the pod's sandbox is created, but also on-demand.
+
+### Thickening multus
+A “thin CNI plugin” runs as a one-shot process, typically as a binary on disk
+that runsrun on a Kubernetes host machine.
+
+A “thick CNI Plugin”, on the other hand, is a CNI component composed of two (or
+more) parts, usually composed of “shim”, and a long lived process (daemon)
+resident in memory. The “shim” is a lightweight “thin CNI plugin” component that
+simply passes CNI parameters (such as JSON configuration, and environment
+variables) to the daemon component, which then processes the CNI request.
+
+To transform multus into a thick plugin, it is needed to instantiate a long
+lived process - which will be the multus pod entrypoint - listening to a unix
+domain socket - this socket must be available both in the multus pod and the
+hosts's mount namespaces; as such, a bind mount to host this socket must be
+provided for the multus pod.
+
+The CNI shim will then be invoked by kubelet, and send the CNI ADD/DELETE
+commands to the server side of multus via the unix domain socket previously
+mentioned.
+
+Refer to the sequence diagram below for more information:
+
+![Create pod flow for thick plugin](create-pod-flow.png)
+
+### Further multus requirements
```

Comment:
One requirement is that there's still a "thin plugin mode" that's available. Tomo noted that this is likely important to edge computing environments where there's limited memory/cpu, and the thick plugin implementation is likely to utilize a greater amount of resources (how much more I'm not sure, part of me thinks it won't be a lot more?)

One possibility here is that we create a `./v3` folder with the current implementation.

I believe we'll likely tag a `v4.0` for the change for thick plugin.

Reply:
I wouldn't like to engage in these talks without having characterized the system.

Having said that, I do think (intuitively) that it will require more memory / CPU than the thin alternative.

I'l add that requirement. Thanks.